_posts/2024-04-09-flake-schemas.md

```diff
@@ -0,0 +1,143 @@
+
+Shown below is the `nix flake show` output of my [NixOS configuration](https://github.com/gvolpe/nix-config) repo:
+
+![schemas](../images/flake-schemas.png)
+
+Neat! Don't you think? I was irked by the usual `homeConfigurations: unknown` output.
+
+## Making it work
+
+One would need to add [DeterminateSystems/flake-schemas](https://github.com/DeterminateSystems/flake-schemas) as an input to `flake.nix`, but you'll find there are a few useful open PRs since last year that solve real problems, but pretty much like the entire feature, they also seem inactive. So to try this out, I've made [a fork](https://github.com/gvolpe/flake-schemas) and applied some of those patches myself.
+
+With that out of the way, we can now define the schema inputs in our flake:
+
+{% highlight nix %}
+{
+  inputs = {
+    nixpkgs.url = "nixpkgs/nixos-unstable";
+    flake-schemas.url = "github:gvolpe/flake-schemas";
+
+    ## nix client with schema support: see https://github.com/NixOS/nix/pull/8892
+    nix-schema = {
+      inputs.flake-schemas.follows = "flake-schemas";
+      url = "github:DeterminateSystems/nix-src/flake-schemas";
+    };
+  };
+}
+{% endhighlight %}
+
+And as outputs, we define the standard schemas plus a custom one we'll see soon:
+
+{% highlight nix %}
+{
+  schemas =
+    inputs.flake-schemas.schemas //
+    import ./lib/schemas.nix { inherit (inputs) flake-schemas; };
+}
+{% endhighlight %}
+
+As pointed out by Greg in [one of the open issues](https://github.com/DeterminateSystems/flake-schemas/issues/14), the `nix` client from the fork doesn't build straight away, so we need some tweaks that the following overlay takes care of:
+
+{% highlight bash %}
+f: p: {
+  nix-schema = inputs.nix-schema.packages.${system}.nix.overrideAttrs (old: {
+    doCheck = false;
+    doInstallCheck = false;
+    postInstall = old.postInstall + ''
+      rm $out/bin/nix-*
+      mv $out/bin/nix $out/bin/nix-schema
+    '';
+  });
+};
+{% endhighlight %}
+
+Then we can add it to our system packages, conveniently named `nix-schema` to avoid conflicts:
+
+{% highlight nix %}
+{
+  environment.systemPackages = [
+    pkgs.nix-schema
+  ];
+}
+{% endhighlight %}
+
+NOTE: the `rm $out/bin/nix-*` bit ensures that all the usual commands such as `nix-build`, `nix-shell`, etc don't get overridden, which is [one of the issues I got myself into](https://discourse.nixos.org/t/the-program-nix-build-is-not-in-your-path/42943).
+
+Next we have the `lib/schemas.nix` file referenced above, which defines the schema for the `out` custom output in my `nix-config` repo, which is a rather simple example of a flake schema:
+
+{% highlight nix %}
+{ flake-schemas }:
+
+{
+  out = {
+    version = 1;
+    doc = "Exports custom attrsets...";
+    inventory = output:
+      flake-schemas.lib.mkChildren (builtins.mapAttrs
+        (_: _: {
+          what = "custom instance to be used by consumers of this flake";
+        })
+        output);
+  };
+}
+{% endhighlight %}
```

Comment:
**suggestion:** frankly I don't understand what this is doing at all myself, so mayyyybe it'd be good to expand on this code sample a little  bit?

Reply:
e.g. what's the goal of the `mapAttrs`, what `mkChildren` does, and more about the standard flake-schemas fields like `version`, `inventory`